### ISSUE TITLE
Some interactions are not recorded in PartonShower in JETSCAPE2.0

### ISSUE THREAD

**Issue by TianyuDai** [JETSCAPE/JETSCAPE#31]
I used examples/brickTest.cc to output the parton shower to test_out.dat. I also printed out all the partons that are pushed back to pOut list directly in the energy loss module (e.g. Martini.cc, Matter.cc). I tried to compare the parton shower in test_out.dat with the parton list directly printed from Martini.cc/Matter.cc. It seems that the parton shower in test_out.dat does not include all the processes happened in Martini/Matter (usually the first event is fine, but the second event starts to have the problem). The final parton list given by examples/FinalStatePartons.cc is also not the same as the final partons that I printed from Martini/Matter. I didn't come across such a problem in previous versions of JETSCAPE. I am not sure if it is a bug or a feature of JETSCAPE2.0. Is there any reason that JETSCAPE2.0 neglect some interactions?

**Comment by TianyuDai** [comment]
I simply print out the parton information before it is pushed back in energy loss module. Then I compare this information with the output file. Here is an example of what I did in Martini.cc: 
```
else if (process == 10)
{
pOut.push_back(Photon(0, 22, 0, pVec, xVec));
pOut[pOut.size()-1].set_form_time(0.);
pOut[pOut.size()-1].set_jet_v(velocity_jet); // use initial jet velocity
fout << pOut[pOut.size()-1].time() << " " << process << " " << pOut[pOut.size()-1].e() << "\n"; 
return; 
}
```

**Comment by j-f-paquet** [comment]
Hello,

Has anybody followed up on this? Is there some additional information we should provide to move the ticket forward?

Thank you,

J-F

**Comment by chunshen1987** [comment]
Do you add a liquefier to the energy loss module? Can you send us a small example of the problem? Thanks!

**Comment by jdmulligan** [comment]
Hi @TianyuDai @j-f-paquet,
Apologies for the delay in getting to this -- will take a look. To clarify: You are running `brickTest` out of the box, without any changes besides the print statement?

**Comment by jdmulligan** [comment]
Forwarding the reply from Tianyu:

> Hi James, 
> 
> Thank you for replying us. I have attached the files that I used when I did the test. In brickTest.cc, I changed the Matter module to Martini module. In Martini.cc, I added the print statement to print the parton information before it is pushed back to pOut. I did not make any other changes. 
> 
> The files named "test_out.dat" and "parton_list" are output files (you can regenerate these files by running brickTest). "test_out.dat" is the output file of brickTest. "parton_list" includes the information of partons that are pushed back to pOut in "src/jet/Martini.cc". The columns in "parton_list" are time, process, parton id and energy of the parton. In my understanding, the parton showering list in "test_out.dat" should include all kinds of interactions. However, when I compare "test_out.dat" with "parton_list", only a few interactions in "parton_list" are recorded in "test_out.dat". Do you know why do other interactions are not recorded? 
> 
> Thank you, 
> 
> Tianyu

[Tianyu_files.zip](https://github.com/JETSCAPE/JETSCAPE/files/3787199/Tianyu_files.zip)

**Comment by jdmulligan** [comment]
Hi @chunshen1987, All,

I reproduced @TianyuDai's observations by running brickTest with Martini enabled in place of Matter. Given Tianyu's printouts and some further digging one immediately sees the issue is that the parton shower object is only being added to when there are >1 partons returned by the energy loss module's `DoEnergyLoss()` function. This was indeed not the case before v2.0. 

I ran the same brickTest on two Jetscape versions:
- Public master
- v1.4

One can quickly see from the output that in v2.0 every entry in the parton shower is from a 1->2 splitting, whereas in v1.4 there are many 1->1 entries. 

In particular, the code change was introduced in `JetEnergyLoss::DoShower()` (commit: https://github.com/JETSCAPE/JETSCAPE/commit/0243890dc40a514143b08c265a3ac2987743ff53), where you can see new special treatment for the case of 1 outgoing parton returned by `Martini::DoEnergyLoss()` (i.e. `pOutTemp.size() == 1`). 

From the development history it looks like these changes were to integrate two-stage hydro -- as far as I am aware the behavior change described above was not intentional -- @chunshen1987 please comment and take a look.

Best,
James

**Comment by chunshen1987** [comment]
Hi all,

The modifications in JetEnergyLoss::DoShower() were introduced to make the parton shower graph physically correct. The 1->1 processes from the LBT module correspond to the free-streaming processes. There is no need to create a new vertex in the parton shower. 

For MARTINI, it implicitly drops soft partons below some threshold in parton energy and results in 1->1 processes. These should be 1->2 processes. The framework v2.0 with the liquefier module corrected for such mislabelled processes and added back the missing energy-momentum at those vertices.

**Comment by jdmulligan** [comment]
Closing the issue since this was evidently intended.
@chunshen1987 it is really essential to communicate any changes that break backward compatibility. 
I will add a description to the release notes since this is crucial information.

**Comment by jdmulligan** [comment]
Re-opening due to follow-up by Tianyu:

> Hello
>  James and Chun,
> 
> Thank
>  you for your explanation! It can explain the incompatibility between JETSCAPE 1.4 and JETSCAPE 2.0. However, the problem is that the modification of parton showering output may result in incorrect final state partons.
> 
> As
>  a user of JETSCAPE, I think we are supposed to run brickTest.cc to generate an output file first, and then run FinalStatePartons.cc to read the output file to get final partons. The final partons that we get is only based on the parton showering list in the
>  output file. As you said, the simplification of parton showering in JETSCAPE 2.0 will neglect 1->1 processes (some of those 1->1 process in Martini involve energy loss). If I run Martini without Liquefier, the final partons that I get will be incorrect, since
>  1->1 processes with energy loss are not included. Do I have to run Martini with Liquefier in JETSCAPE 2.0?
> 
> 
> In
>  general, I tried 4 cases: (i) JETSCAPE parton list w/ Liquefier, (ii) my own parton list w/ Liquefier, (iii) JETSCAPE parton list w/o Liquefier, (iv) my own parton list w/o Liquefier. Only (i) and (ii) are compatible, but others are different from each other.
>  If Liquefier is the module aiming to handle soft partons, why does the result of hard partons be affected?

**Comment by jdmulligan** [comment]
And follow-up by Chun:

> Because of the limiting manpower, the new Liquefier implementation in v2.0 was only tested with the MATTER + LBT setting. No test has been done with the MARTINI module. During the development of v2.0, there are a few features such as recoils were being added to the MARTINI module.
> 
> It is very likely that the current JETSCAPE framework does not take into account all the MARTINI’s conventions on the output Parton list. We will need to discuss in details all the possible output scenarios that MARTINI provide and make sure the framework handles them correctly.

@CPark2016

**Comment by chunshen1987** [comment]
Since today's COMP meeting (Nov. 22) is canceled, @TianyuDai Can you let me know how you need the framework to deal with the case when MARTINI has only one parton in the pout list? Do you require the framework to generate a vertex for this process?

For your comparison (i) & (ii) vs. (iv), what are your criteria to determine a parton is soft? Where is the difference in (i) vs. (iv)?

Thanks!

**Comment by TianyuDai** [comment]
@chunshen1987 It is not necessary for the framework to generate a vertex, but I hope the final state parton list generated by FinalStatePartons.cc to be correct (do not neglect energy loss in this process). 
I consider a parton to be soft if its energy is smaller than pcut (set in xml file). By "JETSCAPE parton list", I mean the paton list generated by FinalStatePartons.cc. By "my own parton list", I mean the parton list I directly obtain from Martini.cc, which is also run in JETSCAPE.

**Comment by chunshen1987** [comment]
Thanks for the explanation! The thing I don't understand is that my modifications were only made to the output graph structure. Those partons go through the 1->1 processes are pushed back to the shower list, meaning that their energies and momenta should be updated in the parton shower.

**Comment by TianyuDai** [comment]
@chunshen1987 Yes, the modification is only made to the output graph structure, but it seems that FinalStatePartons.cc reads the output graph structure to return the parton list, so the change of the output graph structure will also change the final parton list that we get from FinalStatePartons.cc.

**Comment by chunshen1987** [comment]
Hi @TianyuDai (cc. @CPark2016),

I asked Amit (@amitkr2410) about how the FinalStatePartons.cc reads the graph and generates to the final parton list. Now, I think I understand where the inconsistency is.

In v2.0, we impose the requirement that energy-momentum needs to be conserved at every vertex in the graph. The Liquefier is used to find the missing energy-momentum at every vertex and treats them as source terms for hydrodynamics. When the liquefier is not attached, the framework will not generate vertex for the 1->1 process. This is inconsistent with the MARTINI convention as it implicitly drops the soft partons. If the last splitting in MARTINI is such a 1 -> 1 process, then the FinalStareParton.cc will read the wrong information as this process is not present in the graph.

To fix this issue when liquefier is not present, one can impose a check for energy and momentum in the framework after every Jet Energy Loss module passes back the parton list. If there are missing energy and momentum, the framework will generate a vertex with the "correct" 1->2 process. However, the hadronization module requires additional parton's status and species information from the graph. It is hard for the framework to figure out that information based on the returned parton list from the individual energy loss module. Therefore, the ideal solution to this problem is to require MARTINI to return the soft parton information to the framework that it drops. In this case, the framework always interprets the 1 -> 1 process as free-streaming. If a parton has energy less than pcut, MARTINI can just free-streaming it instead of applying energy loss to it.

Does this sound like a solution to your problem? Let me know what you think.

Chun

**Comment by CPark2016** [comment]
Hi all,
I think Chun's suggestion is simplest thing to do. 
By the way, I added this feature in release2.0 version pull request in private repository and will be able to do the same thing here once we decide to do.

**Comment by TianyuDai** [comment]
@chunshen1987 @CPark2016 Thank you! I think this is the best solution! Is it possible for you to include this feature in the JETSCAPE 2.2 release?

**Comment by TianyuDai** [comment]
@CPark2016 Hi Chanwook, is the modification that you made in the "release2.0_martini-dev" branch? I think it solves the problem! However, I came across another problem when I run this branch. examples/brickTest.cc works properly, but examples/FinalStatePartons.cc cannot return the parton list. It actually returns nothing. Do you have any idea why does this happen?

**Comment by CPark2016** [comment]
@TianyuDai This is because this branch was cloned before compatibility update was done. If you don't mind using release2.0-* branch, one possible way would be to use release2.0_qmAmit branch. In this branch, all modifications made in release2.0_martini branch are applied. However, please note that the jet cross section in pp could been modified due to HQ implementation.

**Comment by TianyuDai** [comment]
I modified the code as @chunshen1987 and @CPark2016 suggested, and the problem has been solved. However, the code is slowed down significantly because I have to push back a lot of soft particles to the parton list at every time step. As @chunshen1987 mentioned, hadronization module requires these additional parton's status and species information from the graph. However, I'm wondering if we can give only the information of hard recoil particles to the hadronization module? Or is the information of soft particles (~T) also important to the hadronization module?
